enhancements/sha-bundle-digest.md

```diff
@@ -0,0 +1,163 @@
+
+---
+
+title: SHA-Bundle-Digest
+
+authors:
+
+- "@theishshah"
+
+reviewers:
+
+- TBD
+
+approvers:
+
+- TBD
+
+creation-date: 2022-01-14
+
+last-updated: 2022-01-14
+
+status:implementable
+
+---
+
+  
+
+# Title
+
+  
+
+Add support for SHA-based Digest for Images
+
+  
+
+## Summary
+
+  
+
+Currently, operators generated by operator-sdk will store and retrieve container images based on the image name and the image tag in the bundle. This EP adds support to Operator-SDK such that images can be stored and retrieved in the bundle with an image name and an image digest as an optional replacement for a tag. Image digests are SHA-256 sums that are hashed from the image layer therefore uniquely identify one specific image.
+
+  
+
+## Motivation
+
+  
+
+This set of changes will allow SDK users in disconnected environments to generate bundles with the image digest
+
+  
+
+Users in air gapped environments are currently unable to get images based on tags, and can only access local registries based on the digest
+
+This change is necessary due to non-deterministic behavior with images identified via tags, this issue specifically occurs in the OCP registry due to how the digest is generated with tag name as inputs
+
+  
+
+### Goals
+
+  
+  
+
+-   Give users the ability to the enable digest-based image resolution via a CLI flag when generating bundles.
+    
+-   Allow user to specify additional images to resolve digests via an environment variable.
+    
+-   Add ability to resolve image digests from image tag dynamically using local docker registry.
+    
+-   OPEN QUESTION: Store digest and tag mappings in file so that user-written code can resolve tags to digest.
+    
+
+  
+
+### Non-Goals
+
+  
+
+-   Migrate existing Operator-SDK-based operators from image tags to image digests.
+    
+-   Identify locations in user code that reference images by tag.
+    
+
+  
+  
+
+## Proposal
+
+-   Add `--image-digests` flag to `make bundle` in order to enable this feature to be run. TODO(ish): Add example command being run with this flag.
+    
+-   During bundle generation, discover all environment variables that begin with `RELATED_IMAGE_` then resolve the tags to digests. These variables should be set in `config/manager/manager.yaml` as suggested in [https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions). 
+    
+
+```
+ - name: RELATED_IMAGE_che_server_secure_exposer_jwt_proxy_image
+
+   value: quay.io/eclipse/che-jwtproxy:0.10.0
+
+- name: RELATED_IMAGE_single_host_gateway
+
+  value: quay.io/eclipse/che--traefik:v2.5.0-eb30f9f09a65cee1fab5ef9c64cb4ec91b800dc3fdd738d62a9d4334f0114683
+```
+
+-   Generate OCP cluster CSV to allow the docker daemon to resolve image tags to digests. 
+    
+
+  
+  
```

Comment:
Just whitespace issues here.

```suggestion
## Proposal

-   Add `--image-digests` flag to `make bundle` in order to enable this feature to be run. TODO(ish): Add example command being run with this flag.

-   During bundle generation, discover all environment variables that begin with `RELATED_IMAGE_` then resolve the tags to digests. These variables should be set in `config/manager/manager.yaml` as suggested in [https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions). 

```
 - name: RELATED_IMAGE_che_server_secure_exposer_jwt_proxy_image
   value: quay.io/eclipse/che-jwtproxy:0.10.0
- name: RELATED_IMAGE_single_host_gateway
  value: quay.io/eclipse/che--traefik:v2.5.0-eb30f9f09a65cee1fab5ef9c64cb4ec91b800dc3fdd738d62a9d4334f0114683
```

-   Generate OCP cluster CSV to allow the docker daemon to resolve image tags to digests. 
```

Reply:
I don't see anything stating that we will generate the `CSV.spec.relatedImages` field. Are we not doing that? @tlwu2013

Reply:
> spec.relatedImages (user): a list of image tags containing SHA digests mapped to in-CSV names that your Operator might require to perform their functions. 

I'd vote for generating the`CSV.spec.relatedImages` field (see [in the doc](https://sdk.operatorframework.io/docs/olm-integration/generation/#csv-fields)) since it requires SHA digests too. This would help a lot rather than having users fill that manually.

BTW, regarding those example links, it'll be better to include the links to the original file so readers can easily see the raw file to get a better idea.

e.g. Taking a community Operator, eclipse-che Operator, as an example:

**1.** Author needs to add `RELATED_IMAGE_*` in `config/manager/manager.yaml` ([see example](https://github.com/eclipse-che/che-operator/blob/main/config/manager/manager.yaml#L86-L87)).

**2.** `make bundle --image-digests` generates CSV as:
-->
**2-1.** generate SHA ref. in CSV's deployment's env var ([see in example CSV yaml](https://github.com/eclipse-che/che-operator/blob/main/bundle/stable/eclipse-che-preview-openshift/manifests/che-operator.clusterserviceversion.yaml#L1125-L1126)) 

**2-2.** generate CSV's `spec.relatedImage` from deployment's env var ([see in example CSV yaml](https://github.com/eclipse-che/che-operator/blob/main/bundle/stable/eclipse-che-preview-openshift/manifests/che-operator.clusterserviceversion.yaml#L1604-L1606))